### PR TITLE
separating finite and infinite 3d planes

### LINKED ISSUE #12388: Adding a `FlatSurface` or `Rect3d`
## What problem does this solve or what need does it fill?
First of there are many names which describe this structure such as finite plane, bounded plane, flat surface or rect(3d). I'll go with rect3d for this post since it is the most intuitive to me.

Currently, there is no primitive buildin to express a rect in 3d space.
There are however usecases, such as:
* Using the rect3d to for raycasting/collision detection (imagine looking through a portal)
* Using the rect3d to describe a surface, such as a water surface
* Using the rect3d to create meshes.

The current alternative would be the [`Plane3d`](https://docs.rs/bevy/0.13.0/bevy/math/prelude/struct.Plane3d.html) which is describes a unbounded plane. Not quite what we need her but close.
I also believe that separating the abstraction of a infinite and finite plane is necessary to avoid confusion and unwanted sideeffects down the line. 
This topic has already been raised by @alice-i-cecile in this issue https://github.com/bevyengine/bevy/issues/12243

## What solution would you like?
A struct `Rect3d` 
implementing `Meshable` and being a Bounding volume

The actual name of the struct is open to debate

### PR DESCRIPTION
**Pull request by andristarr**
# Objective

Fixes #12388

## Solution

- Removing the plane3d and adding rect3d primitive mesh

### CODE REVIEW

**Comment by NthTensor** [comment]
Nice work! Need to rename the existing `Rectangle` primitive to `Rect2d` I think. Thoughts on adding `dir2` to `Rect3d`? With transforms it's in for a penny in for a pound.

**Comment by andristarr** [comment]
So in the end the decision was to use Plane3d for a finite and InfinitePlane3d for an infinite plane, so all the confusing naming convention with the rectagnle is gone now :)

File: crates/bevy_math/src/primitives/dim3.rs
Comment (NthTensor) [inline crates/bevy_math/src/primitives/dim3.rs:115]: Use of "infinite" in error for finite plane.

Comment (andristarr) [reply crates/bevy_math/src/primitives/dim3.rs]: Good catch, this is fixed now.

**Comment by andristarr** [comment]
This should be ready for final review

**Comment by andristarr** [comment]
I suspect this is the base of the upcoming work to  have infinite plane meshes?

**Comment by NthTensor** [comment]
Yes, I'll try to review this soon. I've had less time for reviews lately.

File: crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs
Comment (NthTensor) [inline crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs:24]: This seems wrong. Why would an *infinite* plane have bounds?

File: crates/bevy_math/src/primitives/dim3.rs
Comment (NthTensor) [inline crates/bevy_math/src/primitives/dim3.rs:73]: I would be concerned about changing the semantics of a public type like this, but this just changes it back to what it was before right?

File: crates/bevy_math/src/primitives/dim3.rs
Comment (NthTensor) [inline crates/bevy_math/src/primitives/dim3.rs:160]:
```suggestion
    pub fn new(normal: TryInto<Dir3>) -> Self {
        Self {
            normal: normal.try_into().expect("normal must be nonzero and finite"),
        }
    }
```

File: crates/bevy_render/src/mesh/primitives/dim3/plane.rs
Comment (NthTensor) [inline crates/bevy_render/src/mesh/primitives/dim3/plane.rs:11]: If this just wraps `Plane3d`, can't we remove it?

**Review by NthTensor — CHANGES_REQUESTED** [review]
Thanks for your patience, I know this took me a while to get to. There's a few things I think we need to fix and double check before this is ready for merge, but it's a big improvement already, and it's in very good shape generally.

Great job!

File: crates/bevy_render/src/mesh/primitives/dim3/plane.rs
Comment (Jondolf) [reply crates/bevy_render/src/mesh/primitives/dim3/plane.rs:11]: Right now, we could, but we should probably add subdivisions back in a follow-up, so it'll be needed for that

File: crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs
Comment (Jondolf) [reply crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs:24]: Semantically I agree, but it can still be useful to be able to compute the AABB for collision detection purposes. You would still want planes (and half-spaces) to be included in BVHs for example. Or you could have a `Shape3d` trait that requires `Bounded3d`.

For reference, Parry's [`HalfSpace` has AABB methods](https://github.com/dimforge/parry/blob/8823198d704a85b748b8c45ee2f54e35f24a080e/src/bounding_volume/aabb_halfspace.rs)

File: crates/bevy_math/src/primitives/dim3.rs
Comment (Jondolf) [reply crates/bevy_math/src/primitives/dim3.rs:73]: The original `Plane` in `bevy_render` was bounded (but had no configurable normal and was always facing +Y), but the `Plane3d` primitive was always unbounded.

If we want to split it into a finite plane and and an infinite plane though, then I think this semantic change is kinda unavoidable, unless we instead name this something like `FinitePlane3d` or `BoundedPlane3d`, but that might be less user-friendly

File: crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs
Comment (NthTensor) [reply crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs:24]: I'll grant that, and it makes more sense looking at the actual impl. I consider this resolved.

File: crates/bevy_math/src/primitives/dim3.rs
Comment (NthTensor) [reply crates/bevy_math/src/primitives/dim3.rs:73]: Agreed. We just need to highlight it really clearly in the migration notes.

File: crates/bevy_render/src/mesh/primitives/dim3/plane.rs
Comment (NthTensor) [reply crates/bevy_render/src/mesh/primitives/dim3/plane.rs:11]: Sounds good.

File: crates/bevy_math/src/primitives/dim3.rs
Comment (andristarr) [reply crates/bevy_math/src/primitives/dim3.rs:73]: I think it there is no code change required by this one then, correct?

File: crates/bevy_math/src/primitives/dim3.rs
Comment (andristarr) [reply crates/bevy_math/src/primitives/dim3.rs:160]: Sorry, the suggested change here produces an error, I am curious what is the driver behind this suggestion?

I think other methods also expect a Vec3 so I wouldn't change just one's signature.

**Comment by andristarr** [comment]
Thanks for reviewing! I have resolved the merge conflict as well, just awaiting your thoughts on my replies! :)

File: crates/bevy_math/src/primitives/dim3.rs
Comment (NthTensor) [reply crates/bevy_math/src/primitives/dim3.rs:73]: Correct

File: crates/bevy_math/src/primitives/dim3.rs
Comment (NthTensor) [reply crates/bevy_math/src/primitives/dim3.rs:160]: If you have a `Dir3` this will do a redundant conversion to and from a `Vec`. It looks to me like this is the only place a vector is taken as a parameter and immediately converted into a `Dir3`.

It's fine either way. What's the error?

Comment (NthTensor) [reply crates/bevy_math/src/primitives/dim3.rs:160]: Ah, it should be `impl TryInto<Dir3>`. Well anyway it's not super important.

**Review by NthTensor — APPROVED** [review]
Looks like my feedback last week wasn't super helpful. Sorry about that, and thanks for the clarifications.

Happy to approve this now, looks good.

**Comment by andristarr** [comment]
This should be ready for merge @alice-i-cecile

**Comment by NthTensor** [comment]
I think we need another approval first. Maybe @Jondolf can give it a look?

File: crates/bevy_math/src/primitives/dim3.rs
Comment (alice-i-cecile) [reply crates/bevy_math/src/primitives/dim3.rs:160]:
```suggestion
    pub fn new(normal: impl TryInto<Dir3>) -> Self {
        Self {
            normal: Dir3::new(normal).expect("normal must be nonzero and finite"),
        }
    }
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I think the TryInto API is meaningfully nicer: I've left a corrected suggestion on how to do this. Once that's done though I'm happy to merge :)

File: crates/bevy_math/src/primitives/dim3.rs
Comment (andristarr) [reply crates/bevy_math/src/primitives/dim3.rs:160]: @NthTensor Sorry, I missed your comments here. Thank you for the explanation!

Comment (andristarr) [reply crates/bevy_math/src/primitives/dim3.rs]: @alice-i-cecile  This is now addressed!

**Comment by NthTensor** [comment]
I'm going to go ahead and mark this as ready for final review, given Alice's conditional approval has been addressed.

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**
